### PR TITLE
Custom Strides for Batched Gemm

### PR DESCRIPTION
**Pull request by OuadiElfarouki**
Goal : Extend batched gemm implementation to allow setting custom strides between batch matrices. 

This patch introduces  3 additional argument parameters (A,B and C matrices strides) with no defaults to the `_gemm_batched`. 

Both `strided` and `interleaved` batched gemm variants are supported, with a restricted set of stride values fixed per case when using interleaved layout _(Check updated Documentation)_. 

All relevant tests have been added and/or updated.

### CODE REVIEW

File: src/interface/gemm_interface.hpp [hunk 1]
```diff
@@ -115,22 +119,48 @@ typename sb_handle_t::event_t _gemm_backend(
...
+  // Checking Strides conformity when batch_size>1
+  index_t a_size = _TrA ? _M * _lda : _K * _lda;
+  index_t b_size = _TrB ? _ldb * _K : _N * _ldb;
+  index_t c_size = _ldc * _N;
+
+  if (batch_size > index_t(1)) {
+    if (batch_type == gemm_batch_type_t::strided) {
+      if (_stridec < c_size || _stridec < 0) {
+        throw std::invalid_argument("invalid _stridec");
+      } else if (_stridea < 0) {
```
Comment (muhammad-tanvir-1211) [inline src/interface/gemm_interface.hpp:132]: I think we need to make the if conditions for `stride_a` and `stride_b` a bit more exhaustive. The user can set a stride which is much larger than the `batch size` or the `lda` and I am not sure that the GEMM implementation would be able to cover those cases.

File: src/operations/blas3/gemm_interleaved.hpp [hunk 2]
```diff
@@ -157,8 +158,14 @@ class Gemm<input_t, output_t, /* DoubleBuffer = */ false, /* NbcA = */ false,
   const index_t ldb_;
   const index_t ldc_;
   const index_t batch_size_;
+  index_t stridea_;
+  index_t strideb_;
+  index_t stridec_;
+  // Matrices Strides are of no use interleaved
```
Comment (muhammad-tanvir-1211) [inline src/operations/blas3/gemm_interleaved.hpp:164]: Did you plan on removing this comment or did you put it here intentionally?

File: src/operations/blas3/gemm_local.hpp [hunk 3]
```diff
@@ -138,6 +139,8 @@ class Gemm<input_t, output_t, DoubleBuffer, NbcA, NbcB, ClSize, TileType,
   static_assert(cl_elems % packetize_t::packet_size == 0,
                 "Cache line size must be a multiple of packet_size");
 
+  // TODO : assert strides are multiples of matrices sizes or = 0
```
Comment (muhammad-tanvir-1211) [inline src/operations/blas3/gemm_local.hpp:142]: I don't think we could add these `static_asserts` here as the `strides` are all runtime parameters.

**Review by s-Nick — APPROVED** [review]

File: doc/Gemm.md [hunk 4]
```diff
@@ -115,6 +115,12 @@ All `GEMM` kernels support batched operations but the interleaved `GEMM` can onl
 
 Batched `GEMM` is called with a separate `_gemm_batched` function, however beyond the user facing functions all `GEMM` calls take the same path, with `batch_size` and `batch_type` parameters controlling if and how a batched operation takes place.
 
+The `_gemm_batched` operation assumes all the matrices have the same parameters (sizes and leading dimensions). In the `strided` batch type case, all matrices are stored within a fixed stride-distance provided by the user (`stride_a`, `stride_b` and `stride_c`), the latter must be at least equal to the matrix c size to avoid overlapping writes to the output. In the `interleaved` batch type case, the strides can only take these values depending on the case :
```
[the remark this answers is not in the document]
Comment (OuadiElfarouki) [reply doc/Gemm.md]: Thanks @pgorlani, elevant clarifications have been added.

File: src/operations/blas3/gemm_interleaved.hpp (hunk 2 above)
Comment (OuadiElfarouki) [reply src/operations/blas3/gemm_interleaved.hpp]: Comment removed. Thanks.

File: src/operations/blas3/gemm_local.hpp (hunk 3 above)
Comment (OuadiElfarouki) [reply src/operations/blas3/gemm_local.hpp]: Comment removed as well.

File: src/interface/gemm_interface.hpp (hunk 1 above)
Comment (OuadiElfarouki) [reply src/interface/gemm_interface.hpp:132]: Following the most recent oneMKL's `gemm_batch` [spec](https://spec.oneapi.io/versions/latest/elements/oneMKL/source/domains/blas/gemm_batch.html#onemkl-blas-gemm-batch), there are no conditions on the strides values, as the user is responsible of passing a valid buffer that can be accessed given the batch_size, matrices sizes and the stride values.

**Review by pgorlani — APPROVED** [review]

**Review by s-Nick — APPROVED** [review]

File: src/interface/gemm_interface.hpp [hunk 5]
```diff
@@ -115,22 +119,38 @@ typename sb_handle_t::event_t _gemm_backend(
...
   bool _TrB = _TransB != 'n';
+
+  // Checking Strides conformity when batch_size>1
+  index_t a_size = _TrA ? _M * _lda : _K * _lda;
+  index_t b_size = _TrB ? _ldb * _K : _N * _ldb;
+  index_t c_size = _ldc * _N;
+
+  if (batch_size > index_t(1) && (batch_type == gemm_batch_type_t::strided)) {
+    if (_stridec < c_size || _stridec < 0) {
+      throw std::invalid_argument("invalid _stridec");
```
Comment (muhammad-tanvir-1211) [inline src/interface/gemm_interface.hpp:130]: Change `invalid_argument` with `assert`.

**Review by muhammad-tanvir-1211 — APPROVED** [review]

**Pull request merged**